File: summer-school/MPI/Exercise_03/solution_03.c
#include <stdlib.h>
#include <stdio.h>
#include <mpi.h>

/* #define ndata 1000 */
#define ndata 1000000


int main(int argc, char* argv[]){

    int me, nprocs, i=0, left, right;
    MPI_Status status;
    MPI_Request request; 

    float a[ndata];
    float b[ndata];

    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &nprocs);
    MPI_Comm_rank(MPI_COMM_WORLD, &me);

    /* Initialize data */
    for(i=0;i<ndata;++i)
      {
    a[i] = me;
    b[i] = -1;
      }
    /* Compute neighbour ranks. By using the module operation, we can guarantee periodicity */
    right = (me+1)%nprocs;
    left  = (me-1+nprocs)%nprocs;
    
    /* Sendrecv data */
    MPI_Isend(a, ndata, MPI_FLOAT, right, 0, MPI_COMM_WORLD, &request);
    MPI_Recv(b, ndata, MPI_FLOAT, left, 0, MPI_COMM_WORLD, &status);

    MPI_Wait(&request,&status);

    printf("\tI am task %d and I have received b(0) = %1.2f \n", me, b[0]);

    MPI_Finalize();
    return 0;
}

File: summer-school/MPI/Exercise_03/solution_03.f90
program hello

  use mpi

  implicit none

  integer ierr,me,nprocs,left,right
  integer status(MPI_STATUS_SIZE)
  integer request

  integer,parameter :: ndata = 1000000

  real :: a(ndata)
  real :: b(ndata)

  call MPI_INIT(ierr)
  call MPI_COMM_SIZE(MPI_COMM_WORLD, nprocs, ierr)
  call MPI_COMM_RANK(MPI_COMM_WORLD, me, ierr)
  !$ Initialize data
  a = my_rank
  b = -1
  !$ Compute neighbour ranks. By using the mod operation, periodicity is guaranteed.
  right = mod(me + 1 , nprocs)
  left =  mod(me - 1 + nprocs , nprocs)
  !$ Sendrecv data

  call MPI_ISEND(a,ndata,MPI_REAL,right,0,MPI_COMM_WORLD,request,ierr)
  call MPI_RECV(b,ndata,MPI_REAL,left,0,MPI_COMM_WORLD,status,ierr)

  call MPI_WAIT(request,status,ierr)

  print *,'I am proc ',me,' and I have received b(1) = ',b(1)
  call MPI_FINALIZE(ierr)

end program hello
